### ISSUE TITLE
Maintaining both GD3 and GD4 Flatpaks?

### ISSUE THREAD

**Issue by Zishan-Rahman** [flathub/org.godotengine.Godot#115]
Hi all.
I don't know if a GitHub issue like this is the best place to ask this but, with the _actual_ release of Godot 4 fast approaching ([right now we're at RC6](https://godotengine.org/article/release-candidate-godot-4-0-rc-6/)) and introducing things that will break existing Godot 3 projects, I was wondering whether or not you folks would maintain Godot 3 **and** Godot 4 Flatpaks when the latter finally reaches a stable release (and how that would be done if so). I personally think it would be a great idea, but it's not up to me, and I completely understand if it would be too much hassle to maintain 2 Godot Flatpaks at once.

**Comment by Calinou** [comment]
This should be feasible using the multiple branch system, but the discoverability in GUI applications and on flathub.org is poor.

I wonder if it makes more sense to offer two different Flatpak entries instead, similar to https://flathub.org/apps/details/org.supertuxproject.SuperTux vs https://flathub.org/apps/details/org.supertux.SuperTux-Milestone1.

**Comment by TingPing** [comment]
You should make a different package with a new app-id such as `org.godotengine.Godot3`:

Pros:

- Users can install both in parallel easily. Using a branch only one launcher can be active at a time.
- Both versions are discoverable in app stores.
- Both store independent settings and will never conflict.

Cons:

- You have to make a new package and submit it.
- Anywhere the app-id is referenced must change.

**Comment by TingPing** [comment]
If you only plan on supporting 3 for a short period, like months. You could keep 4 in the beta branch until 4.1 releases.

**Comment by Calinou** [comment]
Thanks for the advice :slightly_smiling_face: I'll go for the "new package" approach.

**Comment by CharlieQLe** [comment]
My PR should be updated to Godot 4.0 at some point today, so that can be used as a base for a new package.

**Comment by mossymossmoss** [comment]
Any update on this?

**Comment by Calinou** [comment]
I'm away next week, so @CharlieQLe or someone else will have to create a pull request that updates the manifest to Godot 4, but without adding .NET support. This is because (last time I checked), using a .NET-enabled editor will require exported projects to contain the .NET runtime, even if you only use GDScript in a project.

**Comment by Zishan-Rahman** [comment]
> I'm away next week, so @CharlieQLe or someone else will have to create a pull request that updates this repository's manifest to Godot 4, but _without_ adding .NET support.
> 
> This is because (last time I checked), using a .NET-enabled editor will require exported projects to contain the .NET runtime, even if you only use GDScript in a project.

@Calinou [I just opened a new pull request for this. It's my first time working with Flatpaks like this, but I'd like to help somehow.](https://github.com/flathub/org.godotengine.Godot/pull/116)

**Comment by Zishan-Rahman** [comment]
[So now that we actually have Godot 4.0 in this Flatpak build](https://github.com/flathub/org.godotengine.Godot/commit/3c038ba71725efed2d62186e0ae6c1aa75e7acf2), I'm guessing we're still planning on maintaining a separate Godot 3 Flatpak for a while? If so, how do we do it and when do we start? Just curious :eyes:

**Comment by CharlieQLe** [comment]
> This is because (last time I checked), using a .NET-enabled editor will require exported projects to contain the .NET runtime, even if you only use GDScript in a project.

Unrelated to the issue, though related to your comment, could bundling both the non-C# and the C# editors in the Flatpak work? It may be a bit much since C# support would entail building Godot a second time, and that some users have no need for C# support, but it would only be a temporary solution until the unified editor arrives.

**Comment by Calinou** [comment]
> [So now that we actually have Godot 4.0 in this Flatpak build](https://github.com/flathub/org.godotengine.Godot/commit/3c038ba71725efed2d62186e0ae6c1aa75e7acf2), I'm guessing we're still planning on maintaining a separate Godot 3 Flatpak for a while? If so, how do we do it and when do we start? Just curious :eyes:

I think someone needs to create an issue on https://github.com/flathub/flathub/ to request `org.godotengine.Godot3` to be created. (You have my approval for this if you need it :slightly_smiling_face:)

**Comment by ssokolow** [comment]
For anyone landing here while searching for a workaround, in the mean time, you can prevent Godot from upgrading with `flatpak mask org.godotengine.Godot`.

**Comment by reisaraujo-miguel** [comment]
I think it would be better to have a `org.godotengine.Godot-LTS` or something, so when Godot 5 eventually comes around you can just upgrade `org.godotengine.Godot` to Godot 5 and `org.godotengine.Godot-LTS` to Godot 4.

**Comment by Zishan-Rahman** [comment]
> I think it would be better to have a `org.godotengine.Godot-LTS` or something, so when Godot 5 eventually comes around you can just upgrade `org.godotengine.Godot` to Godot 5 and `org.godotengine.Godot-LTS` to Godot 4.

@reisaraujo-miguel This is also a pretty good idea imo. I just created a new issue on Flathub and just added your proposal to it as well.

**Comment by Zishan-Rahman** [comment]
It's been 9 days, so I decided to create a Godot 3 Flatpak manifest using my fork of this Flatpak. [Here it is](https://github.com/Zishan-Rahman/org.godotengine.Godot/tree/Godot3).

Unfortunately, I can't build the Godot 3 Flatpak on my system, because the glu module in the shared-modules submodule isn't being built, [and it's an issue that's being worked on as I write this](https://github.com/flathub/shared-modules/pull/239). Nonetheless, the Flatpak should be able to build on Flathub's CI/CD as it did prior to the 4.0 update.

@Calinou If I could get your approval to open a new PR on the Flathub repository, I'd very much appreciate it. [I already forked the repository with the new Godot 3 Flatpak in a submodule so as not to erase any prior commit history](https://github.com/Zishan-Rahman/flathub/tree/Godot3).

**Comment by Calinou** [comment]
> @Calinou If I could get your approval to open a new PR on the Flathub repository, I'd very much appreciate it

Sounds good :slightly_smiling_face:

**Comment by Zishan-Rahman** [comment]
> > @Calinou If I could get your approval to open a new PR on the Flathub repository, I'd very much appreciate it
> 
> Sounds good :slightly_smiling_face:

Thanks for your approval! I'll get to it when I can.

**Comment by Zishan-Rahman** [comment]
> I already forked the repository with the new Godot 3 Flatpak in a submodule so as not to erase any prior commit history.

[I figured out how to copy the commit history over to my Flathub repository so it's no longer a submodule](https://github.com/flathub/flathub/pull/3988#issuecomment-1471693833), and since I have a properly defined top-level manifest in there now, we shouldn't be faced with any more problems pushing it over to Flathub (or so I hope).

**Comment by Zishan-Rahman** [comment]
The Godot 3 Flatpak is almost complete, but right now I'm facing an issue with flatpak-spawn access (which the Flatpak will need for external editor support). [I left a comment here basically going over my issues and essentially asking for help](https://github.com/flathub/flathub/pull/3988#issuecomment-1474525787), and if anyone here can help, post a comment under that pull request. I'd very much appreciate it. In the meantime, [there's a test build of the Godot 3 Flatpak that you can try out](https://github.com/flathub/flathub/pull/3988#issuecomment-1474437597). [Instructions on how to add flatpak-spawn access manually are in the README](https://github.com/flathub/flathub/blob/3aa82b01fd0fe17fb77d02bbbeb11ec273d6f911/README.md#using-an-external-script-editor-read-warning).

**Comment by Zishan-Rahman** [comment]
[We finally got a Godot 3 Flatpak up on Flathub now](https://flathub.org/apps/details/org.godotengine.Godot3). The flatpak-spawn issues I had before have now been fixed, so you should be able to use external editors with _that_ Flatpak as you would with _this_ one. We can close this issue now.
